APP-5519: Migrated `AtlanClient._default_client` to thread-local storage + caches now bound to optional client

Aryamanz29 opened the pull request:
This PR fixes `pyatlan` caching issues in concurrent environments:
https://atlanhq.atlassian.net/browse/APP-5519?focusedCommentId=367914

References for why we opted for `TLS` to solve this:
https://docs.python.org/3/library/threading.html
https://superfastpython.com/thread-local-data
https://stackoverflow.com/a/105025
https://stackoverflow.com/questions/1408171/thread-local-storage-in-python

Aryamanz29 commented:
> I'm a bit confused. I thought we were getting rid of the get_default_client method on the AtlanClient class.
Also since the caches have a reference to the the client and the client has a reference to the caches, I don't know why we need the get_cache or refresh_cache class methods Seems to me that the client should just lazily create an instance of the cache as needed. Am I missing something?

Yes @ErnestoLoma - in the last standup, we all agreed to completely remove `get_default_client`. Since it's a `ClassVar`, it causes issues in concurrent/multi-threaded environments, as it's shared across threads, affecting its value and cache logic, leading to inconsistent behavior reported by users.  

I tried removing all instances of `get_default_client` and the `cache_key` logic in `get_cache` methods. However, I got stuck in `AtlanTagName` due to Pydantic pre-validators.  

For example, the `Purpose` asset has an `AtlanTagName` field, and its pre-validator converts tag hash strings into human-readable display text. Since pre-validators run before instance creation, they call `AtlanTagCache` to resolve tag hashes, requiring a client to make API calls. But the client is an instance variable (`AtlanTagName(client)`), making it inaccessible to class methods and pre-validators.  

And if we make the client as a `ClassVar` to fix this ^^ then would be equivalent to using `get_default_client()`, which defeats the purpose of per-cache client management. Instead, I propose making the client optional in the constructor while ensuring the default client remains thread-safe using TLS. This approach maintains SDK backward compatibility ✅

-----

Most of the cache inconsistency issues reported by users (and which I was able to replicate) stem from `_default_client` being a `ClassVar`, making client switches non-thread-safe.  

You might wonder why there are so many client switches now <-- this is due to the recent `JWT` token timeout change in production. We now expire API tokens after `15` minutes, and the SDK's token refresh mechanism creates a new client after expiration, leading to a new cache key.  

Since `caches (dict)` is also a `ClassVar` in different cache classes, this further compounds the issue. However, now that we've migrated both `_default_client` and `caches (dict)` from `ClassVar`  to `TLS`, we're safe to use our `cache_key` logic again, which also helps reduce redundant API calls to populate caches.  

I've also tested these changes across multiple scenarios (switching clients, multiple clients isolated in threads), and everything works as expected without introducing any breaking changes. Hopefully, this will fix most issues, but we’ll see what new challenges arise after releasing this change 💪  - optimistic for positive results!  🤞 🚀

-----

```py
  @classmethod
    def __get_validators__(cls):
        yield cls._convert_to_display_text
   
   @classmethod
    def _convert_to_display_text(cls, data):
        from pyatlan.cache.atlan_tag_cache import AtlanTagCache

        if isinstance(data, AtlanTagName):
            return data
        if display_text := AtlanTagCache.get_name_for_id(data):
            return AtlanTagName(display_text)
        else:
            return cls.get_deleted_sentinel()
 ```

```py
      purpose_asset = {
        "typeName": "Purpose",
        "attributes": {
            # AtlanTagName
            "purposeClassifications": [
                "some-hash-string1",
                "some-hash-string1",
            ],
        },
    # Verify that deleted tags are correctly set to `None`
    # assert purpose.atlan_tags == [AtlanTagName('(DELETED)')]
    assert purpose.atlan_tags and len(purpose.atlan_tags) == 2
    assert purpose.atlan_tags[0].type_name.__repr__() == f"AtlanTagName('{DELETED_}')"
    assert purpose.atlan_tags[1].type_name.__repr__() == f"AtlanTagName('{DELETED_}')"
    assert purpose.purpose_atlan_tags and len(purpose.purpose_atlan_tags) == 2
    assert purpose.purpose_atlan_tags[0].__repr__() == f"AtlanTagName('{DELETED_}')"
    assert purpose.purpose_atlan_tags[1].__repr__() == f"AtlanTagName('{DELETED_}')"
```

ErnestoLoma commented:
> I tried removing all instances of `get_default_client` and the `cache_key` logic in `get_cache` methods. However, I got stuck in `AtlanTagName` due to Pydantic pre-validators.

@Aryamanz29 so if the problem is limited to the _convert_to_display_text function then the solution should focus there and not in the AtlanTagCache. If we get  `_default_client` in that block then we can retrieve the cache from the client. Plus there's a fundamental problem in the way we are setting the setting `_default_client` as the following block of code demonstrates:
```
import os

from pyatlan.client.atlan import AtlanClient
from pyatlan.model.assets import (
    Table,
)

client = AtlanClient(base_url="https://fieldtec41.atlan.com")
client2 = AtlanClient(base_url="https://field-sandbox.atlan.com", api_key=os.environ["OTHER_API_KEY"])
table = client.asset.get_by_guid(guid="47835445-20c4-438f-af23-c4f0f8f491e2", asset_type=Table)
print(table.atlan_tags)

```
If we run this code it fails because we are retrieving the asset from fieldtec41 but when the attempt is made to get the display text for the tag the _default_client is set to field-sandbox so a check for the tag is made in field sandbox and the tag doesn't exist there. If the line starting with client2 is commented out the code runs fine. 
Currently we are setting `_default_client` when a client is created, instead what we should be setting `_default_client` to `self` whenever a method of `AtlanClient` is invoked. 

The validator could then be changed to the following:
```
 @classmethod
  def _convert_to_display_text(cls, data):
        from pyatlan.client.atlan import AtlanClient

        if isinstance(data, AtlanTagName):
            return data
        if display_text := AtlanClient.get_default_client().atlan_tag_cache.get_name_for_id(data):
            return AtlanTagName(display_text)
        else:
            return cls.get_deleted_sentinel()
```
The name of the variable should also be changed from _default_client to _current_client to better reflect what it is. 
I believe this would eliminate the need for the class methods on the various caches as well as a need for the cache_key since each client has it's own set of caches.
 Also why do we create a new client when the token is refreshed instead of just updating the associated `Session` header?
cc: @cmgrote

ErnestoLoma reviewed (changes requested):
See comments in conversation.

Aryamanz29 commented:
Thanks @ErnestoLoma for the review!  🙌 

1. Removed `cache_key` maintenance from caches since it's no longer needed. All public cache methods are now instance methods instead of `@classmethod`.  

2. Updated unit and integration tests accordingly and fixed mypy errors. Some tests (`source_tag_cache`, `connection_cache`, redundant mocks) still need refactoring, which I'll address tomorrow — please ignore them for now.

3. Now, `_current_client` is set whenever a user invokes `@property` methods on `AtlanClient` (e.g:`client.user.*`, `client.group.*`, `client.atlan_tag_cache.*`). I've tested your code snippet, and it works as expected ☝️ ✅   

#### Regarding your point about `_default_client`, while we no longer set it in `AtlanClient.__init__`:

>Currently we are setting _default_client when a client is created, instead what we should be setting _default_client to self whenever a method of AtlanClient is invoked.

There is a necessity for cases where no `@property` methods are called e.g:`client.user.*`, `client.group.*`, `client.atlan_tag_cache.*`) - e.g: resolving classification names via `AtlanTagCache`. Without this, `ErrorCode.NO_ATLAN_CLIENT_AVAILABLE` would be raised 👇

_Therefore, we might still need to [set it in the `AtlanClient` constructor ](https://github.com/atlanhq/atlan-python/pull/573/files#diff-500c50175051b3516ef627be78118b1896998f2fb5cdac266dd028c5e58f2c28R225)**+** [in client `@property` methods](https://github.com/atlanhq/atlan-python/pull/573/files#diff-500c50175051b3516ef627be78118b1896998f2fb5cdac266dd028c5e58f2c28R234)  (which, in any case, sets it again and resolves the error in your provided code snippet above ⬆️ )._

---

```py
from pyatlan.client.atlan import AtlanClient
from pyatlan.model.assets import Table, Referenceable

client = AtlanClient()

referenceable = Referenceable()
referenceable.classification_names = [123, "456"]

print(referenceable.atlan_tag_names)
```

```py
class Referenceable
...
...
   @property
    def atlan_tag_names(self) -> List[str]:
        from pyatlan.client.atlan import AtlanClient
        from pyatlan.model.constants import DELETED_

       # We need current client here -> AtlanClient.get_current_client() to call `AtlanTagCache`
        if self.classification_names:
            return [
                AtlanClient.get_current_client().atlan_tag_cache.get_name_for_id(tag_id)
                or DELETED_
                for tag_id in self.classification_names
            ]
        return []
 ```
